Client/Application/AppDelegate.swift

```diff
@@ -89,16 +89,31 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     }
 
     func application(application: UIApplication, openURL url: NSURL, sourceApplication: String?, annotation: AnyObject?) -> Bool {
-        if let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false) where components.scheme == "firefox" && components.host == "open-url" {
-            if let query = components.query, item = components.queryItems?.first as? NSURLQueryItem {
-                if item.name == "url" && item.value != nil {
-                    if let newURL = NSURL(string: item.value!) {
-                        let tab = self.tabManager.addTab(request: NSURLRequest(URL: newURL))
-                        self.tabManager.selectTab(tab)
+        let components = NSURLComponents(URL: url, resolvingAgainstBaseURL: false)
+        if components!.scheme != "firefox" && components!.scheme != "firefox-x-callback" {
+            return false
+        }
+        switch components!.host! {
+        case "open-url":
+            if let query = components!.query, items = components!.queryItems as? [NSURLQueryItem] {
+                if items[0].name == "url" && items[0].value != nil {
+                    if let newURL = NSURL(string: items[0].value!.unescape()) {
+                        self.browserViewController.openURLInNewTab(newURL)
+                        return true
+                    }
+                }
+            }
+        case "x-callback-url":
+            if let query = components!.query, items = components!.queryItems as? [NSURLQueryItem] {
+                if items[1].name == "url" && items[1].value != nil {
```

Comment:
This arbitrary indexing into the `components` query is pretty terrible. What's going on here? Why is this index 1 and not 0, like above? At minimum, we should have verbose commenting here explaining the URL formats, but it'd be even better to figure out a more elegant way to handle this.